### PR TITLE
Student Form Design and Validation Complete

### PR DESCRIPTION
**Pull request by danialsyed6**
Created UI for student add/edit form. Applied validation using react hook form.

### CODE REVIEW

File: src/components/studentForm/index.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  const [error, setError] = useState(false);
+
+  const { handleSubmit, control } = useForm({
+    resolver: yupResolver(studentSchema),
+    mode: 'onChange',
+  });
+
+  const onSubmit = (data: any) => {
+    console.log(data);
+  };
```
Comment (ahmadfaraz629) [inline src/components/studentForm/index.tsx:28]: Use proper types for data

File: src/components/dashboard/optionsMenu/index.tsx [hunk 2]
```diff
@@ -3,19 +3,32 @@ import { IconButton, Menu, MenuItem } from '@mui/material';
 import MoreVertIcon from '@mui/icons-material/MoreVert';
 import DeleteIcon from '@mui/icons-material/Delete';
 import CreateIcon from '@mui/icons-material/Create';
+import { useNavigate } from 'react-router-dom';
 
 const ITEM_HEIGHT = 48;
```
Comment (ahmadfaraz629) [inline src/components/dashboard/optionsMenu/index.tsx:8]: tacke this through css or move it to the 
utils/constant.js file

File: src/components/studentForm/index.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import { Button, InputSelect, InputText } from '../common';
+import {
+  gradeOptions,
+  studentSchema,
+  subjectOptions,
+} from '../../state/utils/data';
+
+const StudentForm = () => {
+  const navigate = useNavigate();
+  const formFor = useLocation().pathname.includes('add') ? 'Add' : 'Edit';
```
Comment (ahmadfaraz629) [inline src/components/studentForm/index.tsx:17]: make enum for Add and Edit. 
FormType

File: src/components/dashboard/index.tsx [hunk 4]
```diff
@@ -2,18 +2,21 @@ import React from 'react';
...
   gradesStyle,
   headerStyle,
   headerTextStyle,
 } from './styles';
 import GradeFlag from './gradeFlag';
 import Table from './table';
-import Button from './button';
+import Button from '../common/button';
 import { students } from '../../state/utils/data';
+import { useNavigate } from 'react-router-dom';
```
Comment (ahmadfaraz629) [inline src/components/dashboard/index.tsx:15]: Re arrange the imports.
Built in
components
styles
utils

File: src/components/common/inputSelect.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+const InputSelect = ({ name, label, options, control }: IProps) => {
+  const {
+    field: { onChange, value },
+    fieldState: { error: errorObj },
+  } = useController({ name, control });
+  const [error, setError] = useState(false);
+
+  useEffect(() => {
+    setError(errorObj ? true : false);
+  }, [errorObj, name, value]);
```
Comment (ahmadfaraz629) [inline src/components/common/inputSelect.tsx:29]: No need for this code.
You are introducing a state which will cause this component render twice.

you can use 
const isError = !!errorObj

File: src/components/studentForm/index.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+  const [error, setError] = useState(false);
```
Comment (ahmadfaraz629) [inline src/components/studentForm/index.tsx:19]: Not sure why this state is used here. what kind of error you are expecting.

Read error from hookform directly

Comment (danialsyed6) [reply src/components/studentForm/index.tsx:19]: Some MUI components require a boolean value for the error. That is why this is used.

Comment (alitjanjua) [reply src/components/studentForm/index.tsx:19]: you can make any error boolean using double exclamatory sign i.e.( !! )

File: src/components/studentForm/index.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    console.log(data);
+  };
+
+  return (
+    <Container maxWidth="sm">
+      <Grid container direction="column" rowGap={3} sx={containerStyles}>
+        <Typography
+          component="h1"
+          onClick={() => setError(!error)}
+          sx={titleStyles}
```
Comment (IshfaqAhmedAwan) [inline src/components/studentForm/index.tsx:38]: I could not understand why are we setting this state "setError" on click of typography

File: src/components/studentForm/index.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import { yupResolver } from '@hookform/resolvers/yup';
+import { Button, InputSelect, InputText } from '../common';
+import { containerStyles, titleStyles } from './styles';
+import {
+  gradeOptions,
+  studentSchema,
+  subjectOptions,
+} from '../../state/utils/data';
+import { IStudentForm } from '../../state/ducks/student/types';
+import { FORM_TYPE } from '../../state/utils/enums';
```
Comment (IshfaqAhmedAwan) [inline src/components/studentForm/index.tsx:14]: Again order of import is not followed as asked.

**Review by IshfaqAhmedAwan — CHANGES_REQUESTED** [review]

File: src/components/studentForm/index.tsx (hunk 8 above)
Comment (danialsyed66) [reply src/components/studentForm/index.tsx:14]: Ahmad bhai said to put components first.

File: src/components/studentForm/index.tsx (hunk 7 above)
Comment (danialsyed66) [reply src/components/studentForm/index.tsx:38]: I did it during testing and forgot to remove it.

File: src/components/studentForm/index.tsx (hunk 8 above)
Comment (IshfaqAhmedAwan) [reply src/components/studentForm/index.tsx]: ok

File: src/App.tsx [hunk 9]
```diff
@@ -2,13 +2,15 @@ import React from 'react';
...
-import { DashboardContainer } from './containers';
+import { DashboardContainer, StudentFormContainer } from './containers';
 
 function App() {
   return (
     <BrowserRouter>
       <Routes>
         <Route path="/" element={<DashboardContainer />} />
+        <Route path="/addStudent" element={<StudentFormContainer />} />
+        <Route path="/editStudent/:id" element={<StudentFormContainer />} />
```
Comment (ahmadfaraz629) [inline src/App.tsx:13]: change the routes
/student
/student/:id

File: src/components/common/inputText.tsx [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  OutlinedInput,
+} from '@mui/material';
+import ErrorIcon from '@mui/icons-material/Error';
+import { Control, useController } from 'react-hook-form';
+
+import { formControlStyles } from './styles';
+import { IStudentForm } from '../../state/ducks/student/types';
+
+interface IProps {
+  name: 'name' | 'marks';
```
Comment (ahmadfaraz629) [inline src/components/common/inputText.tsx:16]: why are we using name and marks as a type ? instead of string

Comment (danialsyed6) [reply src/components/common/inputText.tsx:16]: the name was defined by Yup schema so it was only taking 4 values 'name' | 'marks' | 'grade' | 'subject'

**Review by ahmadfaraz629 — APPROVED** [review]
LGTM 👍

**Review by IshfaqAhmedAwan — APPROVED** [review]

**Pull request merged**